### PR TITLE
triton 2.0 changes

### PR DESCRIPTION
**Pull request by kashif**
## What does this PR do?
Fixes triton to work with version 2.0.0.

TODOs:
- [x] Move the syntax to triton2
- [x] Fix fused dropout
- [ ] Fix the blocksparse op API having changed
- [x] Fix fused linear layer
- [x] Update the benchmarks

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
having a look ASAP, thanks a bunch @kashif !

File: experimental/requirements.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
 torch>=1.10
-triton==2.0.0.dev20220206
+triton==2.0.0
```
Comment (blefaudeux) [inline experimental/requirements.txt:2]: hmm this does not actually work, as of now at least (this pip package does not exist yet). Do you have a plan here ? Sync with @ptillet for the release ?

**Comment by blefaudeux** [comment]
@ptillet, how stable is triton 2 these days ? Is it stable enough (when compared to the latest of the 1.x tree) so that we would move all the repo onto 2 ?

**Comment by blefaudeux** [comment]
@kashif I'm seeing quite a few other things to fix (when running `pytest tests -k triton` for instance, issues in the blocksparse and dropout layers). I can fix that and do a PR to your branch if you want, as you prefer !

**Comment by blefaudeux** [comment]
the interface also changed a fair bit around blocksparse, no more support for key padding and attention masking

**Comment by ptillet** [comment]
@blefaudeux  I would say it's actually more stable than Triton 1.x. A bunch of bugs were fixed. Triton 2.0 is also more powerful, so being able to represent more programs also means running into more bugs :D

**Comment by blefaudeux** [comment]
> @blefaudeux I would say it's actually more stable than Triton 1.x. A bunch of bugs were fixed. Triton 2.0 is also more powerful, so being able to represent more programs also means running into more bugs :D

I think that it settles it then, because this PR is about supporting the same programs as before (with triton 1), so this part should be an improvement and is a no brainer for me ! We need to change a fair bit more code though, I'll sync with @kashif

**Comment by blefaudeux** [comment]
It will also automagically bring in bf16 I presume, so definite net win (cf. https://github.com/facebookresearch/xformers/issues/231)

**Comment by blefaudeux** [comment]
> @kashif I'm seeing quite a few other things to fix (when running `pytest tests -k triton` for instance, issues in the blocksparse and dropout layers). I can fix that and do a PR to your branch if you want, as you prefer !

some updates in https://github.com/facebookresearch/xformers/tree/triton-2, the blocksparse API / triton is not fixed, dropout now runs but there's a subtle bug that the unit test caught (it's dropping the same values across all rows ! in general it does not look statistically sound, something must have changed). I can finish this tomorrow if you'd like, or fill free to pick it up, anything works and thanks already @kashif !

**Comment by kashif** [comment]
> @kashif I'm seeing quite a few other things to fix (when running `pytest tests -k triton` for instance, issues in the blocksparse and dropout layers). I can fix that and do a PR to your branch if you want, as you prefer !

Sure let me run this and have a look. Thank you for the quick response!

**Comment by kashif** [comment]
> @kashif I'm seeing quite a few other things to fix (when running `pytest tests -k triton` for instance, issues in the blocksparse and dropout layers). I can fix that and do a PR to your branch if you want, as you prefer !

I believe my issue is that I do not have a modern GPU to test things 😢  Let me see if I can run things remotely

**Comment by kashif** [comment]
@blefaudeux unfortunately I could not figure out why the dropout is not working... I will need to spend more time on it... so feel free to do it if you have an idea...

**Comment by blefaudeux** [comment]
> @blefaudeux unfortunately I could not figure out why the dropout is not working... I will need to spend more time on it... so feel free to do it if you have an idea... 

No problem ! I know this code well (wrote it 😬), it's a bit tricky in that it was optimizing perf for Triton1 (using for instance the fact that you can generate 4 random buffers at once), may not apply well to triton2. I can have a look later this week, thanks a lot for quick starting this already !

**Comment by blefaudeux** [comment]
ok, dropout is fixed, I think that we just need to update the interface for the blocksparse attention to tritonv 2 and then good to go :)

**Comment by kashif** [comment]
@blefaudeux ok cool having a look at that! Thanks!

**Comment by blefaudeux** [comment]
> @blefaudeux ok cool having a look at that! Thanks!

I think that blocksparse + attention mask is still shaky, but the rest should be good :)

**Comment by blefaudeux** [comment]
There's probably something to be done to update the layernorm code also, it works right now but it seems that it's not that fast in fp16+BW, I kind of remember that @ptillet updated that on the triton side

**Comment by kashif** [comment]
@blefaudeux didnt get time today to look... i will try my best tonight!

**Comment by ptillet** [comment]
Actually, I have just realized that I have yet to upstream these layer norm changes I've just talked about :D Will do that today hopefully. It's a pretty good use case of how to use cache eviction hints.

**Comment by blefaudeux** [comment]
I can create a new PR and put you as co-author @kashif if that's ok, I'm fixing a CI problem but should be good to go with the public triton-2 branch by now

**Comment by kashif** [comment]
Thanks

**Comment by blefaudeux** [comment]
closing this and starting a new one @kashif, you'll be co-author of course, but it's easier for me to commit directly to the branch and there were a lot of issues with the CI

**Pull request closed without merging**